File: src/main/java/me/centralhardware/znatoki/telegram/statistic/telegram/callbackHandler/statistic/TimeDeleteCallback.java
package me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.statistic;

import lombok.RequiredArgsConstructor;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.PaymentMapper;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.ServiceMapper;
import me.centralhardware.znatoki.telegram.statistic.telegram.bulider.InlineKeyboardBuilder;
import me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.CallbackHandler;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageReplyMarkup;
import org.telegram.telegrambots.meta.api.objects.CallbackQuery;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.User;

import java.util.UUID;

@Component
@RequiredArgsConstructor
public class TimeDeleteCallback extends CallbackHandler {

    private final ServiceMapper serviceMapper;
    private final PaymentMapper paymentMapper;

    @Override
    public void handle(CallbackQuery callbackQuery, User from, String data) {
        if (!telegramService.isAdmin(from.getId())){
            sender.sendText("Доступ запрещен", from);
        }

        var id = UUID.fromString(data.replace("timeDelete-", ""));

        if (!serviceMapper.getOrgId(id).equals(getTelegramUser(from).getOrganizationId())){
            sender.sendText("Доступ запрещен", from);
            return;
        }

        serviceMapper.setDeleted(id, true);
        paymentMapper.setDeleteByTimeId(id, true);

        var editMessageReplyMarkup = EditMessageReplyMarkup
                .builder()
                .messageId(((Message)callbackQuery.getMessage()).getMessageId())
                .chatId(callbackQuery.getMessage().getChatId())
                .replyMarkup(InlineKeyboardBuilder.create()
                        .row()
                        .button("восстановить", STR."timeRestore-\{id}")
                        .endRow().build())
                .build();
        sender.send(editMessageReplyMarkup);
    }

    @Override
    public boolean isAcceptable(String data) {
        return data.startsWith("timeDelete-");
    }
}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/telegram/callbackHandler/statistic/TimeDeleteCallback.kt
package me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.statistic

import dev.inmo.tgbotapi.Trace
import dev.inmo.tgbotapi.extensions.api.edit.edit
import dev.inmo.tgbotapi.extensions.utils.messageDataCallbackQueryOrNull
import dev.inmo.tgbotapi.extensions.utils.types.buttons.dataButton
import dev.inmo.tgbotapi.extensions.utils.types.buttons.inlineKeyboard
import dev.inmo.tgbotapi.types.queries.callback.DataCallbackQuery
import dev.inmo.tgbotapi.utils.row
import java.util.*
import me.centralhardware.znatoki.telegram.statistic.bot
import me.centralhardware.znatoki.telegram.statistic.mapper.ServiceMapper

suspend fun timeDeleteCallback(query: DataCallbackQuery) {
    val id = UUID.fromString(query.data.replace("timeDelete-", ""))

    Trace.save("deleteTime", mapOf("id" to id.toString()))
    ServiceMapper.setDeleted(id, true)
    val times = ServiceMapper.findById(id)

    val keyboard = inlineKeyboard {
        row { dataButton("восстановить", "timeRestore-$id") }
        if (times.size == 1) {
            if (times.first().forceGroup) {
                row { dataButton("сделать одиночным занятием", "forceGroupRemove-$id") }
            } else {
                row { dataButton("сделать групповым занятием", "forceGroupAdd-$id") }
            }
        }
    }

    query.messageDataCallbackQueryOrNull()?.message?.let { bot.edit(it, replyMarkup = keyboard) }
}
